### PR TITLE
Exclude node_modules

### LINKED ISSUE #665: Exclude node_modules and bower_components from Analysis
`node_modules` is the location where npm stores dependencies, same for `bower_components` for bower.

The contents of neither should be included in the analysis.

[~fabrice.bellingard] has excluded `node_modules` from the npm-supported scanner

### PR DESCRIPTION
**Pull request by inverno**
Fixes #665

### CODE REVIEW

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java [hunk 1]
```diff
@@ -68,13 +68,17 @@
   public static final String IGNORE_HEADER_COMMENTS = PROPERTY_PREFIX + ".ignoreHeaderComments";
   public static final Boolean IGNORE_HEADER_COMMENTS_DEFAULT_VALUE = true;
 
+  public static final String JAVA_SCRIPT_EXCLUSIONS_KEY = PROPERTY_PREFIX + ".exclusions";
+  public static final String JAVA_SCRIPT_EXCLUSIONS_DEFAULT_VALUE = ".*/node_modules/.*";
+
   public static final Version V6_2 = Version.create(6, 2);
 
   @Override
   public void define(Context context) {
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: MAJOR This method has 101 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS138)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java [hunk 2]
```diff
@@ -68,13 +68,17 @@
   public static final String IGNORE_HEADER_COMMENTS = PROPERTY_PREFIX + ".ignoreHeaderComments";
   public static final Boolean IGNORE_HEADER_COMMENTS_DEFAULT_VALUE = true;
 
+  public static final String JAVA_SCRIPT_EXCLUSIONS_KEY = PROPERTY_PREFIX + ".exclusions";
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: @inverno how about just `EXCLUSIONS_KEY`?

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java [hunk 3]
```diff
@@ -68,13 +68,17 @@
...
+  public static final String JAVA_SCRIPT_EXCLUSIONS_DEFAULT_VALUE = ".*/node_modules/.*";
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: @inverno add bower_components

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java [hunk 4]
```diff
@@ -146,6 +150,16 @@ public void define(Context context) {
         .onQualifiers(Qualifiers.MODULE, Qualifiers.PROJECT)
         .subCategory(GENERAL)
         .category(JAVASCRIPT_CATEGORY)
+        .build(),
+
+      PropertyDefinition.builder(JavaScriptPlugin.JAVA_SCRIPT_EXCLUSIONS_KEY)
+        .multiValues(true)
+        .defaultValue(JAVA_SCRIPT_EXCLUSIONS_DEFAULT_VALUE)
+        .name("JavaScript Exclusions")
+        .description("List of regular expressions used for excluding JS files")
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: @inverno 
```
Comma-delimited list of file path patterns to be excluded from analysis of JavaScript files.
```

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java [hunk 5]
```diff
@@ -37,12 +37,18 @@
...
   }
 
   @Test
   public void should_contain_right_properties_number() throws Exception {
-    assertThat(properties()).hasSize(8);
+    assertThat(properties()).hasSize(9);
+  }
+
+  @Test
+  public void should_initialize_exclusions() throws Exception {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java]: @inverno `should_have_default_for_exclusions`

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java [hunk 6]
```diff
@@ -61,21 +67,21 @@ public void should_have_javascript_as_category_for_properties() throws Exception
   public void count_extensions_for_sonarqube_server_6_0() throws Exception {
     Plugin.Context context = setupContext(SonarRuntimeImpl.forSonarQube(Version.create(6, 0), SonarQubeSide.SERVER));
 
-    assertThat(context.getExtensions()).hasSize(13);
+    assertThat(context.getExtensions()).hasSize(15);
   }
 
   @Test
   public void count_extensions_for_sonarqube_server_6_2() throws Exception {
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java]: @inverno may be we can remove this unit tests

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptExclusionsFileFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+  @Test
+  public void should_exclude_node_modules() throws Exception {
+    MapSettings settings = new MapSettings();
+    settings.setProperty(JavaScriptPlugin.JAVA_SCRIPT_EXCLUSIONS_KEY, JavaScriptPlugin.JAVA_SCRIPT_EXCLUSIONS_DEFAULT_VALUE);
+    JavaScriptExclusionsFileFilter filter = new JavaScriptExclusionsFileFilter(settings);
+    assertThat(filter.accept(inputFile("some_app.js"))).isTrue();
+    assertThat(filter.accept(inputFile("node_modules/some_lib.js"))).isFalse();
+    assertThat(filter.accept(inputFile("node_modules/my_lib_folder/my_lib.js"))).isFalse();
+    assertThat(filter.accept(inputFile("node_modules/sub_module/node_modules/submodult_lib.js"))).isFalse();
```
Comment (vilchik-elena) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptExclusionsFileFilterTest.java]: @inverno `sub_module/node_modules/submodule_lib.js`

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java (hunk 4 above)
Comment (inverno) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: Done, without 'Comma-separated'

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java (hunk 2 above)
Comment (inverno) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: `JS_EXCLUSIONS_KEY`

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java (hunk 6 above)
Comment (inverno) [reply sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java]: I'll try to find out at which moment they became the same

**Comment by vilchik-elena** [comment]
@inverno mention new property in readme

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptExclusionsFileFilterTest.java (hunk 7 above)
Comment (inverno) [reply sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptExclusionsFileFilterTest.java]: Done

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java (hunk 6 above)
Comment (inverno) [reply sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java]: @vilchik-elena they are 'accidentally' the same number, but they refer to different extensions. Better to keep the test for the time being

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java (hunk 5 above)
Comment (inverno) [reply sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/JavaScriptPluginTest.java]: Done

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java (hunk 3 above)
Comment (inverno) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java]: Done

**Pull request merged**